src/pubsub/subscribe.js

```diff
@@ -0,0 +1,67 @@
+'use strict'
+
+const ndjson = require('iterable-ndjson')
+const explain = require('explain-error')
+const bs58 = require('bs58')
+const { Buffer } = require('buffer')
+const { objectToQuery } = require('../lib/querystring')
+const configure = require('../lib/configure')
+const { ok, toIterable } = require('../lib/fetch')
+const SubscriptionTracker = require('./subscription-tracker')
+
+module.exports = configure(({ fetch, apiAddr, apiPath, headers }) => {
+  const subsTracker = SubscriptionTracker.singleton()
+
+  return async (topic, handler, options) => {
+    options = options || {}
+    options.signal = subsTracker.subscribe(topic, handler, options.signal)
+
+    const qs = objectToQuery({
+      arg: topic,
+      discover: options.discover,
+      ...(options.qs || {})
+    })
+
+    const url = `${apiAddr}${apiPath}/pubsub/sub${qs}`
+    let res
+
+    try {
+      res = await ok(fetch(url, {
+        method: 'POST',
+        signal: options.signal,
+        headers: options.headers || headers
+      }))
+    } catch (err) { // Initial subscribe fail, ensure we clean up
+      subsTracker.unsubscribe(topic, handler)
+      throw err
+    }
+
+    // eslint-disable-next-line no-console
+    const onError = options.onError || (err => console.error(err))
+
+    ;(async () => {
+      try {
+        for await (const msg of ndjson(toIterable(res.body))) {
```

Comment:
If body is empty, won't this fail miserably?

Reply:
The body is a stream, it will be empty until a message arrives...not sure I follow!?

Reply:
Fair, understood now. My question appeared because I didn't see any step where the answer is empty (e.g. cause of failure)

Reply:
If there's an error in the stream it'll throw and be caught and we clean up :) If a message cannot be parsed it's also caught but it's a "soft" error - we don't drop the subscription. I think one of go-ipfs or js-ipfs (I forget which) sends an empty object `{}` in the stream initially which fails to be parsed...